File: DP_1.py
# Time Complexity : O(mn)
# Space Complexity : O((m+1)(n+1))
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

import math
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [[0 for j in range(amount + 1)] for i in range(len(coins)+1) ]
        
        for j in range(1, len(dp[0])):
            dp[0][j] = math.inf
            
        for i in range(1, len(dp)):
            for j in range(1, len(dp[0])):
                if j < coins[i-1]:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = min(dp[i-1][j], 1 + dp[i][j - coins[i-1]])
                            
        res = dp[-1][-1]
        
        if res >= math.inf:
            return -1
        
        return res

# Time Complexity : O(n) | n -> number of houses 
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def rob(self, nums: List[int]) -> int:
        if nums is None or len(nums) == 0: return 0
        elif len(nums) == 1: return nums[0]
        
        dp = [0 for _ in range(len(nums))]
        dp[0], dp[1] = nums[0], max(nums[0], nums[1])

        for i in range(2, len(nums)):
            dp[i] = max(dp[i-1], dp[i-2] + nums[i])
            
        return dp[-1]

            

File: DP1.java
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*Initialize 0 row 0 column with 0. Constructing a DP table with values in the first row as infinity. 
If value is greater than the denomination of the coin, copy the contents of the row above it with same column. 
Else, compute the min between top row same column and same row, the coin denomination number positions left of the current position + 1.  
 * */

class Solution {
    public int coinChange(int[] coins, int amount) {
        int m = coins.length;
        int n = amount; 
        int dp[][] = new int[m+1][n+1];
        for(int i=1; i <= n; i++)
        dp[0][i] = 99999;
        for(int i =1; i <=m; i++){
            for(int j =1; j <=n;j++){
                if(coins[i-1] > j)
                dp[i][j] = dp[i-1][j];
                else
                dp[i][j] = Math.min(dp[i-1][j], 1+dp[i][j-coins[i-1]]);
            }
        }
        if(dp[m][n] >=99999)
        return -1; 
        return dp[m][n];
    }
}